keps/sig-node/753-keystone-containers/README.md

```diff
@@ -0,0 +1,815 @@
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+- Pod termination by kubelet if application containers have successfully run to completion and exited.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+- Handling Container startup/shutdown ordering for application and sidecar containers
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+A pod annotation can be specified by users which contains a list of all core application containers in the pod. Depending on the user feedback and how we want to proceed with the sidecar proposal this annotation can be modified or promoted to an API field. 
+
+The annotation is used by kubelet to make a decision on pod termination based on the status of application containers and pod restart policy.
+
+- if all application containers exited successfully and the restart policy of the pod is "never" or "onFailure" then terminate the pod.
+
+- if all app containers are failed and the restart policy is "never" then terminate the pod. 
+If failure logs are concerned in this case (considering the logging container is sidecar) then the sidecar container will only be terminated when all app containers are in the failed state, so it must have captured the failure logs.
+
+The following are existing behaviors just mentioned here for clarity:
+
+- if all application containers exited successfully and the restart policy of the pod is "always" then restart the containers.
+
+- if all app containers are failed and the restart policy is "onfailure", then kubelet will keep trying to restart containers, and no point in terminating the pod
+
+Since container annotations are not supported to be added by an end user, We need to use pod annotation to specify application containers.
+
+There are few options for annotations:
+
+1. Specifying a list of keystone or sidecar containers in one annotation
+   e.g. `alpha.node.kubernetes.io/sidecars:{list of sidecar containers}` or
+        `alpha.node.kubernetes.io/keystone:{list of keystone containers}`
+2. One annotation per container
+   e.g. `alpha.sidecars.kubernetes.io/<container name>:string`
+
+The motivation for one annotation per container is easy translation to api fields later on and it can be used for the sidecar ordering proposals.
+
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
```

Comment:
While this section is technically optional, it should probably be filled out so we can consider how the proposal meets the needs of various scenarios that aren't served by current behaviour.